src/analysis/normed_space/units.lean

```diff
@@ -0,0 +1,55 @@
+/-
+Copyright (c) 2020 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+import analysis.normed_space.basic
+import analysis.specific_limits
+import tactic.noncomm_ring
+
+/-!
+# The group of units of a normed ring
+
+In this file we prove that the group of units of a complete normed ring is an open subset of the
+ring.
+-/
+
+noncomputable theory
+
+variables {α : Type*} [normed_ring α] [complete_space α]
+
+/-- In a normed ring, an element `x` of distance less than `1` from `1` is a unit.  Here we
+construct its `units` structure.  -/
+def perturbation_unit (x : α) (h : ∥x∥ < 1) : units α :=
+{ val := 1 - x,
+  inv := tsum (λ (n : ℕ), x ^ n),
+  val_inv := mul_neg_geom_series x h,
+  inv_val := geom_series_mul_neg x h }
+
+lemma unit_of_near_unit [nonzero α] (x : units α) (y : α)
+  (h : ∥y - x∥ < ∥((x⁻¹:units α):α)∥⁻¹) : is_unit y :=
+begin
+  have hpos : 0 < ∥((x⁻¹:units α):α)∥ := norm_pos x⁻¹,
+  have hrad : ∥((x⁻¹:units α):α) * (x - y)∥ < 1,
+  { calc ∥((x⁻¹:units α):α) * ((x:α) - y)∥
+        ≤ ∥((x⁻¹:units α):α)∥ * ∥(x:α) - y∥           : norm_mul_le x.inv _
+    ... = ∥((x⁻¹:units α):α)∥ * ∥y - x∥               : by rw [←neg_sub, norm_neg]
+    ... < ∥((x⁻¹:units α):α)∥ * ∥((x⁻¹:units α):α)∥⁻¹ : by nlinarith [h, hpos]
```

Comment:
```suggestion
    ... < ∥((x⁻¹:units α):α)∥ * ∥((x⁻¹:units α):α)∥⁻¹ : by nlinarith only [h, hpos]
```
or just `nlinarith`, but this is slightly faster. `linarith` and friends will automatically use everything in the context. I'm glad to see `nlinarith` is already getting some use!

Reply:
I would use it all the time, were it not for exactly the reason you mention.  Am I allowed to use `nlinarith` in mathlib?  Or is it like a non-terminal `simp`, something that I can use in draft code but which should be rewritten for performance reasons in production?

Reply:
Another point is that `by linarith` is more human-readable than `ne_of_gt hpos`.  If I am allowed to do the former, I would prefer to!

Reply:
Maybe what you are saying is that `nlinarith only [***]` and `linarith only [***]` are the right compromise.  We need a `squeeze_nlinarith`!

Reply:
`linarith` is completely fine to use. If you have a ton of inequalities/equalities in the context, and the contradiction follows from only a couple, you might consider `linarith only`. Depending on the order of things in the context this may or may not make a big difference.

`nlinarith` is quadratically worse than `linarith` in the number of hypotheses. `nlinarith only` is more likely to see a noticeable speedup than `linarith only`, which is why I suggest it here. 

I tend to think of it more in terms of effort, instead of readability. mathlib isn't so readable anyway. If the manual proof takes a second longer and saves half a second each compilation over 50k compilations, feels worth it. If you have to think to find the manual proof, less worth it. But everyone weights these things differently and there's no standard. So in summary, do whatever you'd like! Since you've already included the relevant hypotheses in the list here and below, changing to `(n)linarith only [...]` takes zero seconds of thought for nonzero speedup.